src/core/SampleBuffer.cpp

```diff
@@ -1447,3 +1457,153 @@ SampleBuffer::handleState::~handleState()
 {
 	src_delete( m_resamplingData );
 }
+
+
+
+// Fading related functions
+
+float SampleBuffer::fadingVal(f_cnt_t pos, bool isLeft)
+{
+	float relativePos = static_cast<float>(pos)/static_cast<float>(m_frames);
+	return (this->*fadefuncPtr)(relativePos, isLeft);
+}
+
+
+void SampleBuffer::setLeftFader(float lfader)
+{
+	m_leftFader = lfader;
+	checkFadingActive();
+}
+
+
+void SampleBuffer::setRightFader(float rfader)
+{
+	m_rightFader = rfader;
+	checkFadingActive();
+}
+
+
+void SampleBuffer::checkFadingActive()
+{
+	if ((m_leftFader > 0.0) || (m_rightFader < 1.0))
+	{
+		m_fading = true;
+	}
+	else
+	{
+		m_fading = false;
+	}
+}
+
+
+float SampleBuffer::bezierFade(float relativePos, bool isLeft)
+{
+	float amp = 1.0f;
+	if (isLeft)
+	{
+		amp *= m_leftBezierFade.evaluate(relativePos / m_leftFader);
+	}
+	else
+	{
+		amp *= m_rightBezierFade.evaluate(1 - (relativePos - m_rightFader) / (1 - m_rightFader));
+	}
+	return amp;
+}
+
+
+float CubicBezier::cubicBezierLastSegmentP1(float t, float p1, float p2)
+{
+	return 2 * p1 * t * (1 - t) + p2 * pow(t, 2);
+}
+
+
+float CubicBezier::cubicBezierLastSegmentP2(float t, float p1, float p2)
+{
+	return p1 * pow(1 - t, 2) + 2 * p2 * t * (1 - t) + pow(t, 2);
+}
+
+
+float CubicBezier::cubicBezier(float t, float p1, float p2)
+{
+	return (
+		cubicBezierLastSegmentP1(t, p1, p2) * (1 - t)
+		+ cubicBezierLastSegmentP2(t, p1, p2) * t
+	);
+}
+
+
+float CubicBezier::cubicBezierDerivative(float t, float p1, float p2)
+{
+	return (
+		3 * (3 * pow(t, 2) - 4 * t + 1) * p1
+		+ 3 * t * (2 - 3 * t) * p2
+		+ 3 * pow(t, 2)
+	);
+}
+
+
+float CubicBezier::solveXforTNewton(float x, bool useCache=true)
+{
+	int maxIter = 100;
+	float t, tUp, f, fPrime;
+	if (useCache) { t = m_tCache; }
+	else { t = 0.5f; }
+	for (int i = 0; i < maxIter; ++i)
+	{
+		f = cubicBezier(t, m_p1.x(), m_p2.x()) - x;
+		fPrime = cubicBezierDerivative(t, m_p1.x(), m_p2.x());
+		tUp = t - f / fPrime;
+
+		if (abs(tUp - t) < 10e-3) { break; }
+		t = tUp;
+	}
+	if (useCache) { m_tCache = tUp; }
+	return tUp;
+}
+
+
+float CubicBezier::solveXforTBisection(float x)
+{
+	float a = 0.0f;
+	float b = 1.0f;
+	float t, fA, fT;
+
+	int maxIter = 10;
+	for (int i = 0; i < maxIter; ++i)
+	{
+		t = (a + b)/2;
+		fT = cubicBezier(t, m_p1.x(), m_p2.x()) - x;
+		fA = cubicBezier(a, m_p1.x(), m_p2.x()) - x;
+		bool aSign = (fA >= 0) ? true: false;
+		bool tSign = (fT >= 0) ? true: false;
+		if (aSign == tSign) { a = t; }
+		else { b = t; }
```

Comment:
If both `m_p1.x()` and `m_p2.x()` are in `[0, 1]`, It can be proved that the cubic polynomial is non-decreasing. Using that, you can reduce one call to `cubicBezier()`:
```suggestion
		if (cubicBezier(t, m_p1.x(), m_p2.x()) < x) { a = t; }
		else { b = t; }
```